### PR TITLE
WEB: Obfuscating workgroup email addresses to fix Issue #51209

### LINKED ISSUE #51209: WEB: Obfuscate workgroup email addresses
We recently added our workgroup email addresses to our website: https://pandas.pydata.org/about/team.html#workgroups

While this has been useful, and we received relevant emails from people that otherwise wouldn't know how to contact us easily, we also started receiving spam. I'm unsure if spam is generated manually by people ending up in our website, or by bots fetching our email addresses automatically. But in case it's the latter, I think it'd be good to see if we can easily obfuscate the email addresses in the html code.

I guess there are many options, but it comes to my mind that something very easy that could possibly stop some of the spam would be to simply prepend a string to the email addresses in the html, and then remove it via javascript. This won't help with spammers getting our addresses manually, or using scrappers with javascript support like selenium, but with some luck most of the spam comes from simpler bots just fetching the html.

The idea would be that for example if the address is `address@pandas.pydata.org`, the html generated from the markdown is something like `<a href="mailto:noaddress@pandas.pydata.org">noaddress@pandas.pydata.org</a>`, and then we have a simple javascript block that removes the `no` and makes the final html rendered to the user contain the right address.

This is the file where this should be implemented: https://github.com/pandas-dev/pandas/blob/main/web/pandas/about/team.md#-workgroupname-

### PR DESCRIPTION
**Pull request by Kabiirk**
Hi,

As defined by the scope of this Issue #51209 , I have implemented a JavaScript based solution for obfuscating ```workgroup``` emails:

If required, please let me know whether any Tests/Code Checks need to be added/passed for this or if any other change needs to be implemented to close this PR successfully .

Thanks

---
- [x] closes #51209 
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by Kabiirk** [comment]
Hi @mroeschke & @datapythonista ,

Going through the details in ```Code Checks / pre-commit (pull_request)```, it seems there is a trailing whitespace in my code that is causing the issue. There is an extra whitespace at line ```59``` in ```team.md``` (Please ref. screenshot below). Shall I make the change and commit again in the same PR ?

Ref. 
image

**Comment by Kabiirk** [comment]
@github-actions pre-commit

**Comment by Kabiirk** [comment]
Okay,

So the ```Code Checks / pre-commit (pull_request)``` is passing now.
But I am not sure why ```Ubuntu / Data Manager (pull_request)``` action is now failing for this PR. I have gone through the action details/summary and and it seems there is an ```AssertionError``` in this codeblock of ```pandas/util/_test_decorators.py``` :
image

But as per my understanding, this should run if the platform is ```Windows``` (line ```250```), so why is it being executed in an Ubuntu based check ? Maybe there is some gap in my understanding, please clarify & let me know the way forward.

Thanks

**Comment by datapythonista** [comment]
Preview docs at: https://pandas.pydata.org/preview/51266/

**Comment by datapythonista** [comment]
Thanks @Kabiirk, this seems to be working as expected: https://pandas.pydata.org/preview/51266/about/team.html and the CI failures seem unrelated to this PR.

I'd personally avoid having the real addresses in the html as said in the issue. I think this is an improvement, but with the approach I proposed we could avoid scrappers using a regex for the email addresses, not only the ones that extract it from the `a` tag analyzing the dom.

**Comment by Kabiirk** [comment]
Let me see what I can do for this. It can probably be done with ```element.innerHTML``` or ```element.setAttribute```, I just have to figure out a non-ugly way to target separate DOMs after Jinja does it's thing.

I'm making these changes in this PR itself.

**Comment by datapythonista** [comment]
> Alternatively, emails from the source itself would need to have a prefix, which the JS script removes (I don't think that's a good Idea).

Not in the source (the config file), but in the template, literally in the same line where you remove the prefix again via javascript. It's not great to have to do this, but I think it's a much simpler approach than any alternative that prevents the email addresses to be in the html, while they are made available to users.

Thanks @Kabiirk

File: web/pandas/about/team.md
Comment (datapythonista) [inline web/pandas/about/team.md:59]:
```suggestion
              '<a href="mailto:asp.' + email + '">'
```
`asp` for `anti spam prefix`, and the html document will be sent as `document.write('<a href="asp.coc@pandas.pydata.org" ...`

File: web/pandas/about/team.md
Comment (datapythonista) [inline web/pandas/about/team.md:61]: Then here you can have something like `$('a.contact').attr('href') = $('a.contact').attr('href').removepreffix('asp.`)`. (sorry for the free style pseudo-code here, but I think you get the idea.

Actually, if we do this, we don't need to create the other javacript dinamically.

The advantage that I see is that what I'm proposing would be resistent if the html document is processes with a regular expression, while your approach is only resistant if it's processed via extracting the `a` tags from the dom.

What do you think?

Comment (datapythonista) [reply web/pandas/about/team.md:61]: Yep, exactly. Sorry, it was a bit tricky to explain and took some time make my idea understandable. Feel free to implement it the way you think it's simpler and clearer, I was just showing some pseudocode to show better the concept. Also, not sure if `asp.` is the best prefix, also up to you to decide what to use.

Thanks for working on this @Kabiirk

Comment (Kabiirk) [reply web/pandas/about/team.md:61]: No worries :)
Happy to help !

**Comment by Kabiirk** [comment]
Hi,

I have implemented the proposed change. The scraper picks up the wrong email (i.e. with the prefix):
image

While the correct email is visible to the end-user after the```<script></script>``` tag executes:
image

Moreover, I was able to completely eliminate the use of ```workgroup.contact``` from the ```<script></script>``` tags. So even if someone takes a raw dump of the HTML response with ```requests.get(TEAM_URL).content.decode()```, they will not get the email from within the```<script></script>``` itself, at least not without manual effort.

File: web/pandas/about/team.md
Comment (datapythonista) [inline web/pandas/about/team.md:56]: Maybe I'm missing something, but splitting by space and then joining should be the same as replacing, and jinja has a replace filter we can use.

Unrelated, the asp can leave directly in the html, a bit simpler than concatenating it to the variable first, no?

```suggestion
        <a id="{{ workgroup.name|replace(' ', '-') }}" href="mailto:asp.{{ + workgroup.contact }}">asp.{{ workgroup.contact }}</a>
```

File: web/pandas/about/team.md
Comment (datapythonista) [inline web/pandas/about/team.md:60]: A bit simpler:

```suggestion
            mail_tag_element.innerHTML = mail_tag_element.innerHTML.replace(/^asp./, "");
```

**Review by datapythonista — COMMENTED** [review]
Thanks @Kabiirk, added couple of suggestions to simplify the code, but looks great.

File: web/pandas/about/team.md
Comment (Kabiirk) [reply web/pandas/about/team.md:56]: Hi , I have tested this and it works , 

changes suggested seem fine, but we would need to do the same change in script tag as well. I'll suggest that change in a few minutes.

`

File: web/pandas/about/team.md
Comment (Kabiirk) [reply web/pandas/about/team.md:60]: The only risk I see in this is that the regex may give a different answer if ```.``` is removed from our prefix or it picks up string which start with ```asp``` for e.g. :
image

but tbh this works in our case and any problem is unlikely to occur with this because our prefix will always have a ```.```. I am fine with this change as well.

**Comment by Kabiirk** [comment]
I'll make the suggested changes in my local branch & commit again.

**Comment by Kabiirk** [comment]
Hi,

I tested this and it works, have committed the same to PR. Let me know if this is fine.

**Review by datapythonista — APPROVED** [review]
Really nice @Kabiirk, looks good to me.

**Comment by Kabiirk** [comment]
Hi @datapythonista ,

Following up on your approval, let me know if this PR is supposed to be merged by me, or it would be merged by you ?

**Comment by datapythonista** [comment]
Sorry for the delay @Kabiirk. We usually want two core developers to review a PR before it gets merged, to be in the safe side. I think we're both done here, let's see if another core dev can do the second review soon, and when they do, they'll merge. Thanks again for the help with this!

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @Kabiirk

**Comment by Kabiirk** [comment]
Thanks a lot @datapythonista  & @mroeschke 

No worries, thanks for clarifying that.
I understand it can be a challenge to comprehensively look at so many PRs & give in-depth inputs ! Really appreciate your help & suggestions on this PR.

**Pull request merged**
